src/main/java/net/minecraftforge/client/model/generators/BlockModelBuilder.java

```diff
@@ -18,7 +27,282 @@
+         * Sets the root transform.
+         *
+         * @param transformation the transformation to use
+         * @return this builder
+         * @throws NullPointerException if {@code transformation} is {@code null}
+         */
+        public RootTransformBuilder transform(Transformation transformation)
+        {
+            Preconditions.checkNotNull(transformation, "Transformation must not be null");
+            this.translation = transformation.getTranslation();
+            this.leftRotation = transformation.getLeftRotation();
+            this.rightRotation = transformation.getRightRotation();
+            this.scale = transformation.getScale();
+            return this;
+        }
+
+        /**
+         * Sets the origi of the root transform.
+         *
+         * @param origin the origin vector
+         * @return this builder
+         * @throws NullPointerException if {@code origin} is {@code null}
+         */
+        public RootTransformBuilder origin(Vector3f origin)
+        {
+            this.originVec = Preconditions.checkNotNull(origin, "Origin must not be null");
+            this.origin = null;
+            return this;
+        }
+
+        /**
+         * Sets the origin of the root transform.
+         *
+         * @param origin the origin name
+         * @return this builder
+         * @throws NullPointerException if {@code origin} is {@code null}
+         * @throws IllegalArgumentException if {@code origin} is not {@code center}, {@code corner} or {@code opposing-corner}
+         */
+        public RootTransformBuilder origin(Origin origin)
+        {
+            this.origin = Preconditions.checkNotNull(origin, "Origin must not be null");;
+            this.originVec = null;
+            return this;
+        }
+
+        /**
+         * Finish configuring the parent builder
+         * @return the parent block model builder
+         */
+        public BlockModelBuilder end()
+        {
+            return BlockModelBuilder.this;
+        }
+
+        public JsonObject toJson() {
+            // Write the transform to an object
+            JsonObject transform = new JsonObject();
+
+            if (!translation.equals(Vector3f.ZERO))
+            {
+                transform.add("translation", writeVec3(translation));
+            }
+
+            if (!scale.equals(ONE))
+            {
+                transform.add("scale", writeVec3(scale));
+            }
+
+            if (!leftRotation.equals(Quaternion.ONE))
+            {
+                transform.add("rotation", writeQuaternion(leftRotation));
+            }
+
+            if (!rightRotation.equals(Quaternion.ONE))
+            {
+                transform.add("post_rotation", writeQuaternion(rightRotation));
+            }
+
+            if (origin != null)
+            {
+                transform.addProperty("origin", origin.getSerializedName());
+            }
+            else if (originVec != null && originVec != Vector3f.ZERO)
```

Comment:
Comparing against `Vector3f.ZERO` should be using `.equals(Object)`, not a reference equality check.
```suggestion
            else if (originVec != null && !originVec.equals(Vector3f.ZERO))
```